File: leetcode/problems/problem-specific/3.longest-substring-without-repeating-characters/3.longest-substring-without-repeating-characters.go
package main

// * Sliding Window Solution | O(n) Time | O(n) Space

func lengthOfLongestSubstring(s string) int {
	sSize := len(s)
	if sSize <= 1 {
		return sSize
	}

	var exists = struct{}{}
	res, hashSet := 0, make(map[byte]struct{})
	left, right := 0, 0
	for right < sSize {
		if _, ok := hashSet[s[right]]; !ok {
			hashSet[s[right]] = exists
			res = findMax(res, len(hashSet))
			right++
		} else {
			delete(hashSet, s[left])
			left++
		}
	}

	return res
}

func findMax(a, b int) int {
	if a >= b {
		return a
	}

	return b
}

File: leetcode/problems/sessions/2023-05-01_2023-09-16/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start

# * Two Pointers and Sliding Window Solution | O(n) Time | O(n) Space
# * n -> The length of s string


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = right = 0
        res = 0
        s_set = set()
        while right < len(s):
            if s[right] not in s_set:
                s_set.add(s[right])
                res = max(res, len(s_set))
                right += 1
            else:
                s_set.remove(s[left])
                left += 1

        return res


# @lc code=end
